### ISSUE TITLE
[general need] Save pipeline/history information in the database

### ISSUE THREAD

**Issue by LStruber** [populse/populse_mia#263]
In order to keep and be able to show history of files present in the main database (see #262), we need to save pipeline information in the database (indeed, brick tag is not enough alone to retrieve all information of the history, cf. discussion in #236).

In order to do that, we need to create a new collection (at least), let's say COLLECTION_HISTORY, that will contains this information. This collection will minimally need to contain:
- an uuid
- the inputs of the pipeline (dictionnaries)
- the outputs of the pipeline (dictionnaries)
- an xml chain (serialization of the pipeline) 

The main database will then have for each file entry a history field containing the uuid of the corresponding history in the COLLECTION_HISTORY.

This ticket is open to discuss the structure of the COLLECTION_HISTORY and the links between different collection of MIA. For example, if we introduce the link between COLLECTION_HISTORY and COLLECTION_CURRENT (via the uuid of history in a field), do we need to keep the link between COLLECTION_CURRENT and COLLECTION_BRICK, which consists of a field containing the uuid of the last brick that created the file, or do we see the COLLECTION_HISTORY as an intermediate collection which stores all links between the history and corresponding bricks (a field that contains all uuid of bricks involved in the pipeline) ?
I think that in this last case, it may be tricky to determine which brick truly created the file (and we could fall again in resolving conflicts as discussed in #236), so we would need to keep both links in the main database (for each file storing in different fields the uuid of the history and the uuid of the last brick)

**Comment by sapetnioc** [comment]
Each history document is related to the execution of a `Process` instance. We decided to save the process instance. It means saving the `Process` as well as its parameters. We also need to be able to know what are the nodes that have actually been executed. Finally, there is a technical need to be able to quickly find the links between an input (or output) file name and an history document identifier. Due to limitations of populse_db v2, I suggests to start by storing them in two lists (even if it is redundant with process parameters field). This will allow the following query to find all the uuids of the history document that used the file `/somewhere/database/image.nii`:
```
db.filter_documents('history', fields=('uuid',), as_list=True, filter='"/somewhere/database/image.nii" IN input_files')
```

Therefore, here is a suggestion of evolution for history collection fields definition:

- **uuid** (str): identifier of the history document. 
- **process** (str): XML representation of the process
- **parameters** (dict[str, Any]): values of all parameters of the process
- **input_files** (list[Path]): list of all files used for this process
- **output_files** (list[Path]): list of all files produced by this process
- **status** (to be defined): structure to define (a some kind of dict ?) containing all necessary information about the process execution. For instance which node have actually run, execution durations, etc.

**Comment by LStruber** [comment]
If we have all uuid of bricks document contained in the process, isn't that enough ? Each brick document contain:
- Name
- Init (done or not done)
- Init time
- Exec (not or not done)
- Exec time
- All inputs of the brick (files or parameters)
- All outputs of the brick

**Comment by LStruber** [comment]
@servoz I feel like it indeed would be easier

**Comment by sapetnioc** [comment]
I do not really know the existing tables therefore I am thinking as if history collection was self content. But since we decided to allow to change the internal structure in future release, I think hat for now the priority is on usage (such as GUI) rather than internal storage and API. Therefore, if you already have necessary information somewhere why not using it.

However, the next big step will be to support history management in Capsul v3. For that step we may choose to move (or duplicate) some information from "bricks" collection to another collection such as history or another one if it makes sense. I do not like the idea to use brick collection name in Capsul because there is no brick in the project. We have executable, process, pipeline; if we want to use brick, we will have to define how it differs from the other concepts. Of course this just my point of view today, this will have be discussed. But for now the focus for history management is on Mia.

**Comment by LStruber** [comment]
Ok. So for now I will create a new collection history that contains:
- uuid (str): identifier of the history document.
- process (str): XML representation of the process
- brick_list (list[str]): list of bricks uuid

I think that will be enough to display history as we discussed, and we'll think on how to merge/change both collections for v3 after.

**Comment by servoz** [comment]
It is true that we have chosen, in Mia, different names for quite similar concepts.

Tell me if I am wrong but I think that in Capsul a process is an atomic element of computation (a unit operation, for example a Smooth), a class, and a node is the graphical representation of this process (a graphical box representing this process). 

In Mia we choseen the term brick for this graphical representation without making any conceptual difference with what is called process in Capsul.

In Mia the brick is as much the graphical representation (the graphical box) as the calculation machine (the class). In fact it's simple (as always I put myself in the shoes of the average user for whom what's in the brick/node (the process) is of little importance, what he wants above all is to have a brick that allows smoothing, in the library, and to be able to put it in a pipeline), and as always what is simple is reductive.

I have no problem with adopting the terms nodes and processes in the Mia instead of the general term brick.

We could rename the tag/collection Bricks, by Nodes or Processes (here we find again the duality: for the lambda user, the most important thing is the graphical box that matters to him, the node. For an informed user, the box is only a box, the most important is the process...).

I don't know what to choose between Nodes and Processes, one is a simplistic and graphical vision, the other is more precise, not graphical... so I let you choose!!!!

**Comment by sapetnioc** [comment]
If browsing the graph that links file names and history document is often use, I would recommend to find a way to quickly associate a file name with a history document uuid. In populse_db v2, there is no database join nor index on elements that are not directly stored in a field (i.e. elements in list or dict). For instance, given a file name, finding the history uuid that created it will require to iterate over all bricks, find the ones using the file, select the last one (how ?) and then iterate over all history documents to find the one having a match on brick_list. This will be highly inefficient when history becomes big. And this will have to be done to follow each link on the graph.

My suggestion would be to create one or two collections putting together two items:
- path (str): name of a file (or directory ?) generated or used by a process
- history_uuid: uuid of an history document using this path

If we choose only one collection, it would be necessary to add a boolean to indicate if the path is used as input or output. Otherwise we could also create two collections, one for inputs and one for output.

In populse_db v2, primary key can only be a single field but, at least for input files, there is no unique path nor history_uuid. Therefore, we will probably need to add a primary key in v2 even if we do not need it (this will not be the case in v3).

**Comment by servoz** [comment]
> If we choose only one collection, it would be necessary to add a boolean to indicate if the path is used as input or output. Otherwise we could also create two collections, one for inputs and one for output.

I understood your explanation. But I don't know how to choose between a collection with a boulean and two collections. For efficiency, speed, what do you recommend?

**Comment by LStruber** [comment]
> For instance, given a file name, finding the history uuid that created it will require to iterate over all bricks, find the ones using the file, select the last one (how ?) and then iterate over all history documents to find the one having a match on brick_list.

Of course we do not want to iterate over all bricks to find the history uuid. To be clear, for now we have the collection of documents COLLECTION_CURRENT that have a BRICK_TAG which contains the uuid of the last brick that created the file (empty for raw data)

What I'm currently doing is to create a COLLECTION_HISTORY that contains the pipeline xml string and the list of all bricks uuid used in the pipeline. I'm also adding to COLLECTION_CURRENT an HISTORY_TAG that contain the uuid of the corresponding entry in the COLLECTION_HISTORY. With that, for each document we can directly access the history that created it (without looping on bricks tag) and then retrieve all bricks contained in the pipeline.

**Comment by servoz** [comment]
> What I'm currently doing is to create a COLLECTION_HISTORY that contains the pipeline xml string and the list of all bricks uuid used in the pipeline. I'm also adding to COLLECTION_CURRENT an HISTORY_TAG that contain the uuid of the corresponding entry in the COLLECTION_HISTORY. With that, for each document we can directly access the history that created it (without looping on bricks tag) and then retrieve all bricks contained in the pipeline.

Ok so we only need one collection (COLLECTION_HISTORY, in place of COLLECTION_BRICK) and a tag in COLLECTION_CURRENT (the HISTORY_TAG), isn't it ?

**Comment by LStruber** [comment]
I kept the COLLECTION_BRICK just to fastly retrieve the last brick that created a file, but it is indeed questionable

**Comment by LStruber** [comment]
In order to (de)serialize a pipeline into an xml string, I suppose I'll have to use functions in `capsul.pipeline.xml`. However these functions seems to (de)serialize pipelines directly into files and do not allow to retrieve the string. Should I divide these function to allow the possibility to retrieve the string without saving it into a file ? Or there is another way ?

**Comment by denisri** [comment]
We'll modify them to work on file-like objects as well. I'll do that.

**Comment by denisri** [comment]
Yo should now be able to do:
```
from capsul.pipeline import pipeline_tools
import io

buffer = io.BytesIO()
pipeline_tools.save_pipeline(pipeline, buffer, format='xml')
string = buffer.getvalue()
```
Note that the xml format (based on `xml.etree`) seems to only work with byte buffers (`BytesIO`) while the python formats allows unicode buffers (`StringIO`), at least in pyhon3.

**Comment by LStruber** [comment]
Thank you. And how to deserialize pipeline from the string ?

**Comment by denisri** [comment]
> Thank you. And how to deserialize pipeline from the string ?

I'm also modifying get_process_instance(), I'll tell you.

**Comment by denisri** [comment]
> When I try to serialize pipeline with `pipeline_tools.save_pipeline`, i get the following error:
> 
> ```
> Traceback:
>   File "/home/lucas/populse_dev/populse_mia/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py", line 1303, in initialize
>     self.test_init = self.init_pipeline()
>   File "/home/lucas/populse_dev/populse_mia/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py", line 1622, in init_pipeline
>     pipeline_tools.save_pipeline(pipeline, buffer, format='xml')
>   File "/casa/host/src/capsul/master/capsul/pipeline/pipeline_tools.py", line 1237, in save_pipeline
>     writer(pipeline, file)
>   File "/casa/host/src/capsul/master/capsul/pipeline/xml.py", line 574, in save_xml_pipeline
>     gui_node = _write_nodes_positions(pipeline, root)
>   File "/casa/host/src/capsul/master/capsul/pipeline/xml.py", line 526, in _write_nodes_positions
>     node_pos.set('x', six.text_type(pos[0]))
> TypeError: 'QPointF' object does not support indexing
> ```
> 
> Here is the code I use during initialization:
> 
> ```
> buffer = io.BytesIO()
> pipeline_tools.save_pipeline(pipeline, buffer, format='xml')
> pipeline_xml = buffer.getvalue()
> ```

ah crap. it's an unexpected type in the xml save code. A bug. I need to look at it.

**Comment by denisri** [comment]
can you try again please ?

**Comment by LStruber** [comment]
It's better, no error in the conversion. However I can't put the result in a 'FIELD_TYPE_STRING' into the database, and the FIELD_TYPE_BYTES does not exists. I'll work on fixing that:
```
Error during initialisation of the "NoName" pipeline ...!
Traceback:
  File "/home/lucas/populse_dev/populse_mia/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py", line 1303, in initialize
    self.test_init = self.init_pipeline()
  File "/home/lucas/populse_dev/populse_mia/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py", line 1626, in init_pipeline
    {HISTORY_PIPELINE: pipeline_xml})
  File "/casa/host/src/populse/populse_db/master/python/populse_db/database.py", line 572, in set_values
    raise ValueError("The value {0} is invalid for the type {1}".format(value, field_row.field_type))
ValueError: The value b'<pipeline capsul_xml="2.0" name="CustomPipeline"><doc /><process module="mia_processes.bricks.preprocess.spm.spatial_preprocessing.Smooth" name="smooth_1"><set name="in_files" value="[\'/home/lucas/populse_dev/data/projects/data_test_263/data/raw_data/alej170316_test24042018-IRMFonct_+perfusion-2016-03-17083444-4-T13DSENSE-T1TFE-000425_000.nii\']" /><set name="fwhm" value="[6.0, 6.0, 6.0]" /><set name="data_type" value="0" /><set name="implicit_masking" value="False" /><set name="out_prefix" value="\'s\'" /><set name="smoothed_files" value="\'/home/lucas/populse_dev/data/projects/data_test_263/data/derived_data/salej170316_test24042018-IRMFonct_+perfusion-2016-03-17083444-4-T13DSENSE-T1TFE-000425_000.nii\'" /><set name="output_directory" value="\'/home/lucas/populse_dev/data/projects/data_test_263/data/derived_data\'" /><set name="use_mcr" value="True" /><set name="paths" value="[\'/home/lucas/apps/MATLAB/MATLAB_Runtime/v95/toolbox/spm12\']" /><set name="matlab_cmd" value="\'/home/lucas/apps/MATLAB/MATLAB_Runtime/v95/toolbox/spm12/run_spm12.sh /home/lucas/apps/MATLAB/MATLAB_Runtime/v95 script\'" /><set name="mfile" value="True" /><set name="spm_script_file" value="\'/home/lucas/populse_dev/data/projects/data_test_263/scripts/pyscript_smooth_5557ca0a-6f3f-4016-82ca-2cbe2f4251aa.m\'" /></process><link dest="smooth_1.in_files" source="in_files" /><link dest="smoothed_files" source="smooth_1.smoothed_files" /><gui><position name="smooth_1" x="&lt;built-in method x of QPointF object at 0x7f3e14e9f828&gt;" y="&lt;built-in method y of QPointF object at 0x7f3e14e9f828&gt;" /><position name="inputs" x="&lt;built-in method x of QPointF object at 0x7f3e14e9f9e8&gt;" y="&lt;built-in method y of QPointF object at 0x7f3e14e9f9e8&gt;" /><position name="outputs" x="&lt;built-in method x of QPointF object at 0x7f3e14e9f898&gt;" y="&lt;built-in method y of QPointF object at 0x7f3e14e9f898&gt;" /></gui></pipeline>' is invalid for the type string

```

**Comment by LStruber** [comment]
I made it works by decoding the bytes array into an utf-8 string:
```
self.project.session.set_values(
                COLLECTION_HISTORY, history_id,
                {HISTORY_PIPELINE: pipeline_xml.decode("utf-8")})
```
but as it necessary to specifiy the coding (that may change ?), I'm not sure this is a good idea. What do you think ?

**Comment by sapetnioc** [comment]
It would be better to direclty get a `str/unicode` rather than a `byte` when converting the pipeline in XML. Doesn't the `save_pipeline()` support using a `io.StringIO` instead of `io.BytesIO` ? If not, we could add a `save_pipeline_string` method that would do the encoding.

But since we plan to go to 3.0 where pipelines are serialized in JSON (JSON is always in UTF-8), it is really safe to assume that encoding will not change before the code that save the pipelines.

**Comment by sapetnioc** [comment]
@servoz for the question about one or two collections, I do not clearly foresee the impact of the decision on performances. Therefore, I propose to choose the simplest solution: one collection. When creating this collection fields, do not forget to add an index on `path` and `uuid` (using `create_field('path', 'string', index=True)`) this will speed-up graph browsing.

**Comment by LStruber** [comment]
@sapetnioc I think that @denisri answered to your question yesterday:
> It would be better to direclty get a str/unicode rather than a byte when converting the pipeline in XML. Doesn't the save_pipeline() support using a io.StringIO instead of io.BytesIO ? If not, we could add a save_pipeline_string method that would do the encoding.

> Note that the xml format (based on xml.etree) seems to only work with byte buffers (BytesIO) while the python formats allows unicode buffers (StringIO), at least in pyhon3.

**Comment by denisri** [comment]
Yes that's it: `xml.etree` doesn't seem to accept sting/unicode buffers. But as @LStruber said it's just a matter of decoding the bytes buffer into a unicode string. We can consider doing it inside the functions to load/save XML buffers, it's probably possible.

**Comment by servoz** [comment]
> @servoz for the question about one or two collections, I do not clearly foresee the impact of the decision on performances. Therefore, I propose to choose the simplest solution: one collection. When creating this collection fields, do not forget to add an index on path and uuid (using create_field('path', 'string', index=True)) this will speed-up graph browsing.

OK !

**Comment by LStruber** [comment]
@denisri, when calling `create_xml_pipeline(module, name, xml_file)` from a xml_string, I'm not sure how to specify (and also what is stand for ?) the `module` parameter, could you tell me a bit about it ?

**Comment by LStruber** [comment]
Note that if I set `module` to `None` `create_xml_pipeline` fails with following error:
```
File "/home/lucas/projects/populse/populse_mia/python/populse_mia/user_interface/pop_ups.py", line 4967, in __init__
    pipeline = create_xml_pipeline(None, None, pipeline_xml)
  File "/home/lucas/projects/populse/capsul/capsul/pipeline/xml.py", line 251, in create_xml_pipeline
    x = float(gui_child.get('x'))
ValueError: could not convert string to float: '<built-in method x of QPointF object at 0x7fe2ee2d8270>'
```

I'm not sure if it is related or not

**Comment by denisri** [comment]
As far as I remember, the module is the name of the module where the pipeline class will be inserted in.

The ValueError seems to be something else (QPointF.x instead of QPointF.x() I guess) - I'll fix that.

**Comment by denisri** [comment]
Can you can try again please ?

**Comment by LStruber** [comment]
It changes the error but not the line:
```
File "/home/lucas/projects/populse/capsul/capsul/pipeline/xml.py", line 251, in create_xml_pipeline
    x = float(gui_child.get('x')())
TypeError: 'str' object is not callable
```

**Comment by denisri** [comment]
I'll try it better but it will have to wait for an hour or so...

**Comment by LStruber** [comment]
No problem, tell me when !

> As far as I remember, the module is the name of the module where the pipeline class will be inserted in.

What do you mean by "inserted in" ? For example, if a create the pipeline from the xml_string in the `pop_ups.py` module, and then I add it to a `PipelineDeveloperView` as follows:
`self.pipeline_view = PipelineDeveloperView(pipeline, allow_open_controller=True)`
Which will be the module name ?

**Comment by denisri** [comment]
> No problem, tell me when !
> 
> > As far as I remember, the module is the name of the module where the pipeline class will be inserted in.
> 
> What do you mean by "inserted in" ? For example, if a create the pipeline from the xml_string in the `pop_ups.py` module, and then I add it to a `PipelineDeveloperView` as follows: `self.pipeline_view = PipelineDeveloperView(pipeline, allow_open_controller=True)` Which will be the module name ?

Actually we normally don't use directly the function `create_xml_pipeline()`, it is called from within `get_process_instance()`:
```
engine = capsul_engine()
pipeline = engine.get_process_instance(xml_string)
```
and `None` seems valid for the `module` param, since it is called this way from `get_process_instance`.

**Comment by LStruber** [comment]
Thanks

> and None seems valid for the module param, since it is called this way from get_process_instance.

I did not find where it is used with None value as `module` param, I found None only for the `name` param but ok

**Comment by LStruber** [comment]
It is now working, I can deserialize the pipeline from the xml_string, however after adding it a PipelineDeveloperView, I cannot open the node_controller:
```
Traceback (most recent call last):
  File "/home/lucas/projects/populse/capsul/capsul/qt_gui/widgets/pipeline_developer_view.py", line 3128, in openProcessController
    ce = ProcessCompletionEngine.get_completion_engine(process)
  File "/home/lucas/projects/populse/capsul/capsul/attributes/completion_engine.py", line 649, in get_completion_engine
    if 'capsul.engine.module.attributes' in engine._loaded_modules:
ReferenceError: weakly-referenced object no longer exists
```

**Comment by denisri** [comment]
Ow, that's "funny"... Can I have an example of your XML pipeline ?

**Comment by LStruber** [comment]
`<pipeline capsul_xml="2.0" name="CustomPipeline"><doc /><process module="mia_processes.bricks.preprocess.spm.spatial_preprocessing.Smooth" name="smooth_1"><set name="in_files" value="['/home/lucas/projects/mia_projects/data_cevestoc_1/data/raw_data/alej170316_test24042018-IRMFonct_+perfusion-2016-03-17083444-0-T13DSENSE-T1TFE-000425_000.nii']" /><set name="fwhm" value="[6.0, 6.0, 6.0]" /><set name="data_type" value="0" /><set name="implicit_masking" value="False" /><set name="out_prefix" value="'s'" /><set name="smoothed_files" value="'/home/lucas/projects/mia_projects/data_cevestoc_1/data/derived_data/salej170316_test24042018-IRMFonct_+perfusion-2016-03-17083444-0-T13DSENSE-T1TFE-000425_000.nii'" /><set name="output_directory" value="'/home/lucas/projects/mia_projects/data_cevestoc_1/data/derived_data'" /><set name="use_mcr" value="True" /><set name="paths" value="['/APPS/MATLAB/MATLAB_Runtime/v95/toolbox/spm12']" /><set name="matlab_cmd" value="'/APPS/MATLAB/MATLAB_Runtime/v95/toolbox/spm12/run_spm12.sh /APPS/MATLAB/MATLAB_Runtime/v95 script'" /><set name="mfile" value="True" /><set name="spm_script_file" value="'/home/lucas/projects/mia_projects/data_cevestoc_1/scripts/pyscript_smooth_cdd1e4eb-fd2b-4f23-805e-49d1d4653eba.m'" /></process><link source="in_files" dest="smooth_1.in_files" /><link source="smooth_1.smoothed_files" dest="smoothed_files" /><gui><position name="smooth_1" x="90.0" y="-119.0" /><position name="inputs" x="-167.6095716389599" y="-1.0" /><position name="outputs" x="499.4246679185573" y="-119.0" /></gui></pipeline>`

**Comment by denisri** [comment]
At what time does the error occur ?
I can't trigger the problem. I can even load it in populse_mia (load pipeline, then copy/paste your xml string in the pipeline line), and see the pipeline boxes, edit values etc. I haven't tried to actually run it, but if I understand the problem is earlier ?

**Comment by servoz** [comment]
Since, I think, [this commit](https://github.com/populse/capsul/commit/8020c0414a68ac5ac638306da8311d26ed644c2c), Mia crashes if we click on the Bricks tag.

Ex.
- On the populse_mia fix4#263 branch.
- Open a project (be careful, with this branch, the old projects are not compatible anymore, we have to reconvert from mri_conv for example).
- Make for example a  mia_processes smooth.
- Go to the Databrowser and hit the smooth_1 button in the History column.
 
```
Traceback (most recent call last):
  File "/data/Git_Projects/populse_mia/python/populse_mia/user_interface/data_browser/data_browser.py", line 2167, in show_brick_history
    self.brick_history_popup = PopUpShowHistory(
  File "/data/Git_Projects/populse_mia/python/populse_mia/user_interface/pop_ups.py", line 4967, in __init__
    pipeline = engine.get_process_instance(pipeline_xml)
  File "/data/Git_Projects/capsul/capsul/engine/__init__.py", line 341, in get_process_instance
    instance = self.study_config.get_process_instance(process_or_id,
  File "/data/Git_Projects/capsul/capsul/study_config/study_config.py", line 718, in get_process_instance
    return get_process_instance(process_or_id, study_config=self,
  File "/data/Git_Projects/capsul/capsul/study_config/process_instance.py", line 163, in get_process_instance
    return _get_process_instance(process_or_id, study_config=study_config,
  File "/data/Git_Projects/capsul/capsul/study_config/process_instance.py", line 402, in _get_process_instance
    result = create_xml_pipeline(module_name, None, process_or_id)()
  File "/data/Git_Projects/capsul/capsul/pipeline/xml.py", line 251, in create_xml_pipeline
    x = float(gui_child.get('x')())
TypeError: 'str' object is not callable
```

I'm working out the current unit test issues for this branch, please don't add to it :-))))))))))))) !

**Comment by denisri** [comment]
Yes it's already fixed (see earlier)

**Comment by servoz** [comment]
Oh yes !!!
 Indeed, the problem has been solved for capsul, but rather comes from [this morning's commit](https://github.com/populse/populse_mia/commit/0d4ea1b9cfe75682a562a53a5ccfb5439592f0bd) in populse_mia ...

**Comment by LStruber** [comment]
Did you recreate your pipeline (by actually running it in the pipeline_manager_tab) or did you open a former pipeline ? The writing of the xml string was wrong but has been fixed now. It's working on my side

**Comment by servoz** [comment]
Oh OK !
The old databases are no longer valid (this successive obsolescence is going too fast for me!). I have just pushed a new project_8 in the resources for UT and test_brick_history() is working again. Thank you.

On the other hand, taking the example above, if we click on the smooth_1 button:
It seems to me that the delay before displaying the history is much longer than before (I hope it is not dependent on the size of the pipeline, because here it is only a process smooth!). And if we click again on the button, it goes much faster but we see the exception in the stdout:
```
Exception ignored in: <function PipelineDeveloperView.__del__ at 0x7fe6bf05eb80>
Traceback (most recent call last):
  File "/data/Git_Projects/capsul/capsul/qt_gui/widgets/pipeline_developer_view.py", line 2699, in __del__
    self.release_pipeline(delete=True)
  File "/data/Git_Projects/capsul/capsul/qt_gui/widgets/pipeline_developer_view.py", line 2817, in release_pipeline
    self.setScene(None)
RuntimeError: wrapped C/C++ object of type PipelineDeveloperView has been deleted

```

**Comment by LStruber** [comment]
I also get this exception, I need to work on it. It ressemble to an error we observed few months ago, on a ticket about renaming iterated process (if my memories are good)... But the delay is unperceptible on my side... Let's see with bigger pipelines

**Comment by servoz** [comment]
Ok I just checked and I observe this slowness only on the host. When I am in casa_distro I don't observe it. It may be related to [this old ticket](https://github.com/populse/populse_mia/issues/229)!!? 
In a nutshell:  Nipype uses etelemetry to do internal things (check for available updates etc). We normally disconnect this in the main.py of mia [(os.environ['NO_ET'] = "1")](https://github.com/populse/populse_mia/blob/master/python/populse_mia/main.py#L45).
I have always suspected a difference for this  between casa_distro and l'hote. But without any certainty and I never took the time to see on it because it didn't seem to be a priority until now ... maybe we'll have to look at this ticket one day !

**Comment by LStruber** [comment]
> At what time does the error occur ?
> I can't trigger the problem. I can even load it in populse_mia (load pipeline, then copy/paste your xml string in the pipeline line), and see the pipeline boxes, edit values etc. I haven't tried to actually run it, but if I understand the problem is earlier ?

The problem does not occur when building the pipeline (here node controller can be accessed), but only when we show it from history... I'll investigate

**Comment by LStruber** [comment]
**UPDATE ON THE PROGRESS:** (currently in branch fix4#263)
- The new table/collection history has been added to MIA database. For now it contains three column:
    - an uuid (str)
    - the xml chain of the pipeline (str)
    - the list of bricks/process uuid included in the pipeline (list of str)
-  In the main collection, the uuid of the history has been added for each entry
- This allows:
    - To save the pipeline (during initialization) as an xml string in the database and to reconstruct it from this chain when wanted (when clicking on history button in the data browser)
    - To recover inputs/outputs of each bricks of the pipeline via uuid of used bricks (as a reminder the brick/process collection contains all necessary information on the desired brick)

**However:**
Two exceptions occuring in Capsul during this process need to be fixed:
- After clicking on history button in the data browser, the reconstructed pipeline (from xml) does appear but when right clicking on a node, then click on "open the node controller", MIA crashes with the following exception:
Traceback (most recent call last):
```
  File "/home/lucas/projects/populse/capsul/capsul/qt_gui/widgets/pipeline_developer_view.py", line 3128, in openProcessController
    ce = ProcessCompletionEngine.get_completion_engine(process)
  File "/home/lucas/projects/populse/capsul/capsul/attributes/completion_engine.py", line 649, in get_completion_engine
    if 'capsul.engine.module.attributes' in engine._loaded_modules:
ReferenceError: weakly-referenced object no longer exists
```
- When clicking on the history button twice (or more) in a row, an exception occurs :
```
Exception ignored in: <function PipelineDeveloperView.__del__ at 0x7fe6bf05eb80>
Traceback (most recent call last):
  File "/data/Git_Projects/capsul/capsul/qt_gui/widgets/pipeline_developer_view.py", line 2699, in __del__
    self.release_pipeline(delete=True)
  File "/data/Git_Projects/capsul/capsul/qt_gui/widgets/pipeline_developer_view.py", line 2817, in release_pipeline
    self.setScene(None)
RuntimeError: wrapped C/C++ object of type PipelineDeveloperView has been deleted
```
Although these exception occur in Capsul, I cannot ensure that it is not due to an error on MIA side. Anyway, if you have an idea (especially @denisri) on what's going on here, it's welcome !

**Comment by denisri** [comment]
It's probably in Capsul, I'll try to look at it.

**Comment by denisri** [comment]
The first problem (weak ref on `CapsulEngine` object) was in Mia actually, and it's fixed now: it was "just" that pipelines (from xml) were rebuilt using a temporary (and empty) `CpasulEngine`, which is weakly referenced in the pipeline (through `StudyConfig`). It's better to use the existing and configured engine `Config.get_capsul_engine()`.

**Comment by denisri** [comment]
The 2nd problem is more tricky: the `PipelineDeveloperView` widget (C++) is deleted when the `PopUpShowHistory` widget (C++) is deleted, that is when the python `PopUpShowHistory` object is deleted. But there it doesn't seem to delete the python part of `PipelineDeveloperView` at this time, but somewhere later (probably a mechanism in python or sip), thus at tome point the C++ widget is deleted while the python object is still living, and it doesn't seem to be possible to delete it earlier (I tried deleting it in `PopUpShowHistory.__del__` but it doesn't actually delete it.
I can just test in `PipelineDeveloperView.release_pipeline()` if the C++ object is deleted, and do nothing then, but it's not a very clean solution. I'm not sure there is one...

**Comment by servoz** [comment]
Yes, indeed the first problem is fixed (just tested).
For the second, if there is no solution then there is no problem :-))))))... Especially since it seems to do the job.
If it's just stdout pollution then it works as expected, maybe the solution is to sweep things under the rug (I don't like it but if there's no real solution!) and not print on stdout for that specific exception?

**Comment by denisri** [comment]
I did what I told about in capsul, a test (`sip.isdeleted(self)`) so that it doesn't print the message anymore.

**Comment by servoz** [comment]
oh yes !!!! The 2 issues seem fixed now.
Thanks @denisri

**Comment by LStruber** [comment]
1. I have no problem running SpatialPreprocessing pipeline (on fix4#263 branch):
![image](https://user-images.githubusercontent.com/86590799/162997576-798189ed-c29f-4e51-a8c9-873e3beccb0f.png)
2. Yes it is sometimes the case, in files are missing, I'll check if I can fill them from bricks collection
4. I can easily put history window behind main window and I don't get the white space you mentioned. For resizing window elements, you're right it should be possible.
![image](https://user-images.githubusercontent.com/86590799/162998243-e8e4aec2-c2f4-4f3e-a970-185c7a24e93e.png)

**Comment by denisri** [comment]
About workflow execution:
I have also seen such strange things this afternoon. I have done important modifs in soma-workflow, in order to optimize it and allow much larger workflows as it did (we have actually gained maybe 2 orders of magnitude in performance speed). I have pushed the modifs because some people working with large datasets need them, and all tests have passed, but maybe there are still side effects... I think sometimes stdout / stderr files are lost or erased. I must figure this out, and it might be the cause of the problem.

**Comment by LStruber** [comment]
Note that my last bv_maker was done this morning !

**Comment by denisri** [comment]
So the problem probably happened today. It should help me to find out...

**Comment by denisri** [comment]
In the meantime you can reset soma-workflow to changeset a26c5ce40e4d03a560ea4f42bfed0df590f063f0, before I started all modifs. I'lll tell you when things are fixed. Sorry for the inconvenience.

**Comment by denisri** [comment]
@servoz, can you confirm your soma-workflow was on changeset 088f8e42efce6b9efe526fca1f39cb7267c0395a (today 2022/04/12, about noon) ? I'm not sure I haven't fixed the problem yet in this last change, since I don't observe it anymore, and the last commit was actually a fix about files handling. But as it seems to be more or less random, I'll keep testing...

**Comment by LStruber** [comment]
@denisri on my side, I'm 3 commits late on soma-workflow : 507903cd9c89af19fbc7c26cea30e10f3d70bca2
I'll try on last commit

**Comment by LStruber** [comment]
On last commit, Spatial preprocessing still seems to work (no error in soma-workflow monitoring, neither in the status bar at the bottom of the screen). But still the stdout message

**Comment by LStruber** [comment]
@servoz regarding point 2, do we really need to be able to open the controller or every information is contained in the brick table below ?

**Comment by servoz** [comment]
I am very surprised that it works properly on your side.
So I'm going to start the test again and check everything that could cause problems. I will keep you informed very soon

**Comment by denisri** [comment]
> I just noticed that although soma-workflow indicates that there was no error, there is a brick that has not been run (normalize_12_1):
> It was working in my previous run (before updating soma-workflow with bv_maker):

Oh, there is a missing job in the workflow ? huh I really have to check. I could not cause an error yesterday, I will test again today.

**Comment by LStruber** [comment]
New test this morning (same pipeline):
![image](https://user-images.githubusercontent.com/86590799/163129330-0e2fd000-377c-4da8-8054-83d47e30198c.png)

It seems indeed to be "random"

**Comment by denisri** [comment]
OK I finally ran into this case. It seems that all soma-workflow files have suddenly been removed. I'm investigating... As I said you can reset to a previous commit until I find it out and fix it.

**Comment by servoz** [comment]
I just tested again after meticulous checks.
I am on host (not BV container).
Everything is in master branches except mia on fix4#263, and uptodate
capsul: 2947c22846afc19e06dbf1af254dc8653262caa5
somabase: 572686f104dd509f0b084e10d3856580db50c35
somaworkflow: 088f8e42efce6b9efe526fca1f39cb7267c0395a
populse_mia: 6b1dcce
mia_processes: dd90bd869048c23c2d40f135ddf4bc02cc135f51
populse_db: 55c295b9838f9f1616815ff2f0b45640e6740c04

It's funny (I'm not sure that's the right word to use!) I just re-launched the pipeline and nothing worked (except the very first brick (list_duplicate1). Indeed, it seems to be erratic
![Screenshot from 2022-04-13 10-44-46](https://user-images.githubusercontent.com/37933899/163137275-a7f7e5f6-9743-423a-8003-a128241dbfb0.png)

Furthermore, if we check in the DataBrowser, we see that the segmentation and realignment worked because the outputs exist in the database
![Screenshot from 2022-04-13 10-48-31](https://user-images.githubusercontent.com/37933899/163138402-17cdec5f-d428-4c6a-8775-eebd427a04dc.png)

@denisri tell us if there is anything we can do to help.

**Comment by servoz** [comment]
> In the meantime you can reset soma-workflow to changeset a26c5ce40e4d03a560ea4f42bfed0df590f063f0, before I started all modifs.

exactly, I just tested it and it works like a charm with this commit! Thanks

**Comment by denisri** [comment]
OK I think I have something. I'll push a temporary fix to get things working, then I need to find a better solution. I'm just checking that it actually fixes things.

**Comment by servoz** [comment]
> @servoz regarding point 2, do we really need to be able to open the controller or every information is contained in the brick table below ?

I've just tested your latest commits and I think we're getting something really good. Indeed, the 4-2 point is handled , which is really nice for the user.

Regarding point 4-1, currently the history window is always in front of the main window. This is not too disturbing because when we want to consult the history it is to look at this window immediately.  However as we are non modal, the user can want to do something else at the same time and in this case it can be annoying to have this window which remains always in front of the main window. I think that the history window should be able to go behind the main window and also be minimised. What do you think? If you like, I can look at this.

Regarding point 2, currently the brick table below do the job. We see all parameters brick when we click on a brick. It's very nice. 
In fact, this information is totally redundant with that of the controller. So if we can have all the information also in the controller, why not (belt and suspender!). If having them requires a lot of work maybe we could just use the table below? In this case, the user should not be able to open the GUi of the controller in the history mode. What do you think?

Concerning the highlight of the selected brick. Currently it is surrounded by a thin red line. I think we should have something more visual (if there are a lot of bricks in the pipeline I'm afraid it won't be visible at first sight). What do you think about the rendering when we Ctrl + left click on a brick in a pipeline?

Regarding point 3, this is really not a priority because stdout is definitely for us. I don't think users will go to stdout much, we just have to make sure we don't get too wordy to make it not too complicated to consult.

**Comment by LStruber** [comment]
Regarding point 4.1, it will be hard for me to look at it because on my configuration I can put history window behind main window without any problem... See this picture:
![image](https://user-images.githubusercontent.com/86590799/163153006-04bb30b0-638f-4983-b0ec-3adc2e91fe7e.png)

**Comment by denisri** [comment]
OK I have implemented a better fix and pushed it.
To explain: the problem was in a cleanup which removes orphan files not referenced any longer in the soma-workflow database. Before my last week modifs, all files were all living in the same directory, thus we checked every file in this directory, and erased it if it was no longer referenced in the database. However I introduced a change to avoid having too many files in the same dir (filesystems lose performance when several thousands of files are in the same dir, and I was experimenting very large workflows, which means millions of stdin/stdout files). Thus when a directory contains more than a certain number of files (200 here), we generate subrirectories in a recursive way, so we now have a tree when dealing with many jobs/files. Directories are not referenced in the database, and were not expected to be found in the cleaning function, thus it was simply erasing all subdirectories in soma-workflow files directory...
The cleanup could happen when connecting a client, thus when opening the monitoring GUI. In my tests I was opening it before running a workflow (thus it was OK), or was not using it at all (very large workflows cannot be efficiently monitored using the GUI), so I was not experiencing the problem.

**Comment by denisri** [comment]
> > I just noticed that although soma-workflow indicates that there was no error, there is a brick that has not been run (normalize_12_1):
> > It was working in my previous run (before updating soma-workflow with bv_maker):
> 
> Oh, there is a missing job in the workflow ? huh I really have to check. I could not cause an error yesterday, I will test again today.

My fix does not address this problem at all, and I have never seen it, but if you had it, it should still be around. I have no idea what can cause it...

**Comment by denisri** [comment]
@LStruber you're sure you had not disabled a node by mistake ? ;)

**Comment by servoz** [comment]
> OK I have implemented a better fix and pushed it.

Oh yes, it's working fine now! Thanks @denisri

**Comment by servoz** [comment]
> Regarding point 4.1, it will be hard for me to look at it because on my configuration I can put history window behind main window without any problem... See this picture:


It's very strange, I don't have this possibility at all... 
I'll take a look at that. You have an idea of what could cause this. If it's only me that has this working it means that it's not related to the codes but to my configuration... Strange ...

**Comment by denisri** [comment]
> > Regarding point 4.1, it will be hard for me to look at it because on my configuration I can put history window behind main window without any problem... See this picture:
> 
> It's very strange, I don't have this possibility at all... I'll take a look at that. You have an idea of what could cause this. If it's only me that has this working it means that it's not related to the codes but to my configuration... Strange ...

I'd say this could be some window flags set in the Qt widget (like modal / window modal / stay on top etc.), that may not be handled the same way in different window managers. I don't seem to have the problem on my side.

**Comment by servoz** [comment]
> > > Regarding point 4.1, it will be hard for me to look at it because on my configuration I can put history window behind main window without any problem... See this picture:
> > 
>> It's very strange, I don't have this possibility at all... I'll take a look at that. You have an idea of what could cause this. If it's only me that has this working it means that it's not related to the codes but to my configuration... Strange ...

> I'd say this could be some window flags set in the Qt widget (like modal / window modal / stay on top etc.), that may not be handled the same way in different window managers. I don't seem to have the problem on my side.

Ah ok I didn't know ... I was just planning to look at the window flags. If I remember well, every call of setWindowFlags() will completely override the current settings, so we need to set all the flags at once. Ok I am looking at this.

**Comment by LStruber** [comment]
I changed the way to higlight nodes, as @servoz suggested. I'll work on point 2.

**Comment by LStruber** [comment]
I just thought of a case that is not managed with current implementation:
Brick collection is filled with the uuids of workflow jobs, that are (if I am not mistaken) unitary processes (called bricks in MIA).
Then, if the pipeline presented in history is not only composed of unitary processes, but also subpipelines, we do not display informations about subpipeline (when clicked) in the brick table at the bottom.
How do we want to manage this case ? We may be able to retrieve inputs/outputs of the subpipeline iterating through list of bricks but I'm not sure it will be easy (?) and construct a table with these inputs/outputs. Another solution would be to display in history an exploded view of the pipeline without subpipeline, but it may be disturbing in case of large pipelines.

A side effect is also that if we click in the data browser on a brick that is in a subpipeline of the main pipeline, nothing will be highlighted in the displayed history (indeed this highlighting is currently done by comparing pipeline brick names with brick names of the brick collection, but not iteratively).

**Comment by servoz** [comment]
> I changed the way to higlight nodes, as @servoz suggested. I'll work on point 2.

Yes it's nice !

I see one other thing that doesn't fit.

In the history window if we select another brick, I see two problems. One minor, one major.

The width of the columns do not adapt to the new data. For example, still with the prepocessing pipeline, if we click on the `spatial_preprocessing_1_1.realign1` button in the History column. 
![Screenshot from 2022-04-13 14-55-01](https://user-images.githubusercontent.com/37933899/163184728-d0142e39-1795-4cc3-8941-b0ce8a278008.png)
The parameters have columns with appropriate widths, for example see the parameter `fwhm`. 
![Screenshot from 2022-04-13 14-58-40_1](https://user-images.githubusercontent.com/37933899/163185557-d53cd983-95db-4b5e-930e-e93880047df3.png)
If we now click in the history window on the coregister brick, the column corresponding to the `fwhm` parameter now represents the `apply_to_files` parameter and this one does not have an adapted width.
![Screenshot from 2022-04-13 15-00-32_1](https://user-images.githubusercontent.com/37933899/163187055-7fd5f1ef-5378-4220-a831-e680b5e85008.png)
This is a minor issue (comfort) because we can use the mouse to widen the column. Let's do it. We see that the right parameter is displayed.
![Screenshot from 2022-04-13 15-03-13](https://user-images.githubusercontent.com/37933899/163186264-4ee283ae-1e72-4d35-9164-c02f98845bac.png)
 If we now select the realign brick again, it's worse, the name of the tag changes well in the column but the value of the tag is not reset!
![Screenshot from 2022-04-13 15-04-58](https://user-images.githubusercontent.com/37933899/163186535-74048a14-55c6-4680-9aa0-08d73266c9a4.png)
This is a major issue.

**Comment by denisri** [comment]
I see 2 other issues linked to the new history management:
5. a single brick cannot be run any longer: open Mia, open a project, to to the pipeline manager, select a "mia_processes.bricks.proprocess.spm.Smooth" brick. Click on the brick, select "Filter" for its "in_files" parameter and chose an input, then click the Run button -> an error is produced in the terminal, because it's trying to save the process as an XML pipeline, whereas it is not a pipeline.
6. some pipelines cannot be correctly reloaded as XML, which cause Mia to crash when clicking a data history button. It happens on our Morphologist pipeline, and it's a Capsul problem: I have opened a ticket here: populse/capsul#224.

**Comment by servoz** [comment]
> 5. a single brick cannot be run any longer: open Mia, open a project, to to the pipeline manager, select a "mia_processes.bricks.proprocess.spm.Smooth" brick. Click on the brick, select "Filter" for its "in_files" parameter and chose an input, then click the Run button -> an error is produced in the terminal, because it's trying to save the process as an XML pipeline, whereas it is not a pipeline.

I just tested and I don't reproduce. Have you ever pushed a patch?

**Comment by denisri** [comment]
> > 5. a single brick cannot be run any longer: open Mia, open a project, to to the pipeline manager, select a "mia_processes.bricks.proprocess.spm.Smooth" brick. Click on the brick, select "Filter" for its "in_files" parameter and chose an input, then click the Run button -> an error is produced in the terminal, because it's trying to save the process as an XML pipeline, whereas it is not a pipeline.
> 
> I just tested and I don't reproduce. Have you ever pushed a patch?

No... how strange...

**Comment by servoz** [comment]
> > > 5. a single brick cannot be run any longer: open Mia, open a project, to to the pipeline manager, select a "mia_processes.bricks.proprocess.spm.Smooth" brick. Click on the brick, select "Filter" for its "in_files" parameter and chose an input, then click the Run button -> an error is produced in the terminal, because it's trying to save the process as an XML pipeline, whereas it is not a pipeline.

 > >  I just tested and I don't reproduce. Have you ever pushed a patch?

> No... how strange...

Did you still reproduce ?

**Comment by servoz** [comment]
One last thing (perhaps). If we close Mia and a History window is open, it stays alive after mia is closed.
We should close everything when we close mia? (this is also true for the soma-workflow GUI).
This is not a major problem ...

**Comment by LStruber** [comment]
@denisri I also can launch single bricks.

@servoz for the error in the brick table of channel_info and tissues, could you check if error is present in master branch also ? I did not change this part.

**Comment by LStruber** [comment]
@servoz yes, everything should be closed when MIA is closed, I noticed that but I cannot do everything in the same time :)

**Comment by servoz** [comment]
> I cannot do everything in the same time :)

What? this is even more strange :-))))

**Comment by servoz** [comment]
Ok, thanks for all the tests and the suggestions for improvement. Things may have evolved a bit since we did the first ones. I propose to take stock and gather what remains to be done (this ticket is getting a bit long to see everything at a glance!).

**Comment by servoz** [comment]
> @servoz for the error in the brick table of channel_info and tissues, could you check if error is present in master branch also ? I did not change this part.

Yes, it's the same on master. It's an old display issue!

**Comment by servoz** [comment]
> > I cannot do everything in the same time :)

> What? this is even more strange :-))))

more seriously, that's why I'm going to try to summarise what remains to be done, so that we can share the work.

**Comment by LStruber** [comment]
I was too fast for you this time. I would add:
10. Check if remove_orphan functions are still up to date with the new database, as well as (tightly related) the click on "cleanup obsolete items" button

**Comment by LStruber** [comment]
@denisri In my tests I actually did "export unconnected mandatory plugs"

**Comment by servoz** [comment]
> @denisri In my tests I actually did "export unconnected mandatory plugs"

For me it was "export all unconnected plugs"
![Screenshot from 2022-04-13 17-27-13](https://user-images.githubusercontent.com/37933899/163215721-5e2d5735-159a-42b1-925d-601af821d026.png)
.

**Comment by denisri** [comment]
> > @denisri In my tests I actually did "export unconnected mandatory plugs"
> 
> For me it was "export all unconnected plugs" 

Ah, that's it: I was in the "simplified" case where the "pipeline" contains just a single unconnected node (no export at all), where the user runs thus, not a pipeline but a single process. This use case is "just for comfort" but is important for simplicity of use, as most users will use existing processes or pipelines without building anything in them, thus don't expect to have a "pipeline construction" operation to perform manually before running a ready-to-use process.

**Comment by servoz** [comment]
oh I see!!! I have to admit that I always export the plugs ..... Without exporting anything I reproduce!
![Screenshot from 2022-04-13 17-46-24](https://user-images.githubusercontent.com/37933899/163219363-dca4a8f7-319f-4e63-8ba2-711186a673dc.png)
 Indeed "simplified" case is very interesting for user, and it should work!

**Comment by servoz** [comment]
so can we say that:
1. Pipeline launch is erratic issue --> solved in soma-workflow by @denisri: **_Fixed_**
4. Foreground/background issue --> seems to be ok for everybody now (ok for me):  **_Fixed_**
??

**Comment by servoz** [comment]
If Yes, what do we do with 2? We invest time to populate the input/output in the controller or we hide it in history mode and use only the parameter table below the pipeline ????

**Comment by servoz** [comment]
1- Single brick issue in "simple mode" (no export)
2- Some pipeline cannot be reloaded from xml --> @denisri work on a fix in Capsul. In the meantime I will handle the exception in MIA
3- Brick table issues in updating, column size and sometimes wrong data in some colums
4- Handle the case of bricks that are subpipelines (what should we do ?)
5- Test and handle the case of iterated pipelines
6- Check if remove_orphan functions are still up to date with the new database, as well as (tightly related) the click on "cleanup obsolete items" button
7- stdout contains too much information --> let this for later

+ Node controller inputs/outputs should be populated --> Depending on what we decide

**Comment by servoz** [comment]
> > oh I see!!! I have to admit that I always export the plugs ..... Without exporting anything I reproduce!

> Maybe the simplest fix (and probably useful in other situations) is to allow capsul.pipeline_tools.save_pipeline() to write a single process, maybe by temporarily building a small pipeline with one node and all parameters exported, as workflow_from_pipeline() does. I can look at it.
For 6. I have a partial fix, but there are still problems, and I expect similar issues as what I ran into when trying to build a "fake morphologist" pipeline (some nodes activation was not working, probably because some internal parameters default/optional states were not preserved. Thus it may take some time to fix completely.

Ok, so we have:

1-  Single brick issue in "simple mode" (no export) ---> allow capsul.pipeline_tools.save_pipeline() to write a single process (@denisri)
2- Some pipeline cannot be reloaded from xml --> @denisri work on a fix in Capsul. In the meantime I will handle the exception in MIA
3- Brick table issues in updating, column size and sometimes wrong data in some colums
4- Handle the case of bricks that are subpipelines (what should we do ?)
5- Test and handle the case of iterated pipelines
6- Check if remove_orphan functions are still up to date with the new database, as well as (tightly related) the click on "cleanup obsolete items" button
7- stdout contains too much information --> let this for later

@denisri, please can you gieve your sentiment for the " Node controller inputs/outputs should be populated" point.

Taking in the order of the menu, if it suits you @LStruber  and if you haven't started working on it I can take point 3. (Brick table issues in updating, column size and sometimes wrong data in some colums) ?

**Comment by denisri** [comment]
> > oh I see!!! I have to admit that I always export the plugs ..... Without exporting anything I reproduce!
> 
> Maybe the simplest fix (and probably useful in other situations) is to allow `capsul.pipeline_tools.save_pipeline()` to write a single process, maybe by temporarily building a small pipeline with one node and all parameters exported, as `workflow_from_pipeline()` does. I can look at it.

OK this is fixed in capsul.

**Comment by servoz** [comment]
Wonderful, the list is shrinking as fast as it was built!
1- Some pipeline cannot be reloaded from xml --> @denisri work on a fix in Capsul. In the meantime I will handle the exception in MIA
2- Brick table issues in updating, column size and sometimes wrong data in some colums
3- Handle the case of bricks that are subpipelines (what should we do ?)
4- Test and handle the case of iterated pipelines
5- Check if remove_orphan functions are still up to date with the new database, as well as (tightly related) the click on "cleanup obsolete items" button
6- stdout contains too much information --> let this for later

**Comment by servoz** [comment]
8- Node controller inputs/outputs should be populated  --> let this for later

**Comment by LStruber** [comment]
I take point 3. In short, my idea is to:
- if the selected brick in data browser is inside a subpipeline node:
    - find in which subpipeline resides the brick (iterative function as there can be several nested subpipelines) --> **Issue here:** if several bricks in several subpipelines has the same name this can lead to conflict, (how) can we resolve it ?
    - highlight the subpipeline node
    - display inputs/outputs of the selected brick in the brick table below
- if the user click on a subpipeline node in the history window:
    - highlight the subpipeline node
    - display in the brick table a table that list inputs and outputs of the subpipeline

What do you think ?

**Comment by LStruber** [comment]
@denisri @servoz maybe somewhere in Capsul/MIA there are functions that have already been coded to do what I suggested ?
- from a brick name, find the subpipeline(s) it belongs ?
- from an input/output plug of a pipeline, find the first unitary process it crosses (either forward or backward) ?

**Comment by denisri** [comment]
The brick name cannot be enough to find out the subpipeline: for instance in our Morphologist pipeline, there are two "normalization" subpipelines which are applied first on the raw data, then on the skull-stripped data, and they thus contain the same sub-processes (and sub-pipelines), so the bricks names are not unique.
I think that when we get UUIDs for individual bricks (which used to be done during init, maybe things have changed now you have reworked many things?), we should either reference in the bricks the subpipelines they belong, or (the other way) mark subpipelines or processes with the UUIDs of "bricks". Here I mean "process" or "pipeline" in the pipelining way (an iterated process exists only once) and "brick"/UUID in the workflow way (one UUID for each process execution, thus several for an iteration) - I don't know which is the best way but this has to be recorded somewhere.

**Comment by LStruber** [comment]
In fact (I forgot that when writing my previous message), the name of the bricks in the database is constructed as follows:
`name_of_subpipeline_1.name_of_subpipeline_2.[...].name_of_brick`

So from an "database" brick, we can find out which subpipeline it belongs.

**Comment by LStruber** [comment]
@denisri I just pushed a commit that should correct the crash you reported yesterday

**Comment by servoz** [comment]
> I take point 3. In short, my idea is to:
    if the selected brick in data browser is inside a subpipeline node:
        find in which subpipeline resides the brick (iterative function as there can be several nested subpipelines) --> Issue here: if several bricks in several subpipelines has the same name this can lead to conflict, (how) can we resolve it ?
        highlight the subpipeline node
        display inputs/outputs of the selected brick in the brick table below
    if the user click on a subpipeline node in the history window:
        highlight the subpipeline node
        display in the brick table a table that list inputs and outputs of the subpipeline --> to retrieve inputs/outputs files of the subpipeline, we will need to follow each input/output plug to find the first brick they are linked to, and retrieve in the database associated brick inputs/outputs files. Here again we are facing the same issue than above, we may find several corresponding bricks in the database...
What do you think ?

This seems to me a good approach. 
Since the brick name is of type `sub-pipeline_name_1.sub-pipeline_name_2.[...].brick_name` in the database, does this solve all the problems of determining the brick and sub-pipeline name or am I wrong?
The use of the UUID allows to manage the case where the bricks have the same name, isn't it?

**Comment by LStruber** [comment]
> I take point 3. In short, my idea is to:
> 
>     * if the selected brick in data browser is inside a subpipeline node:
>       
>       * find in which subpipeline resides the brick (iterative function as there can be several nested subpipelines) --> **Issue here:** if several bricks in several subpipelines has the same name this can lead to conflict, (how) can we resolve it ?

This was in fact easy since (as I said after) the brick_name contains the subpipeline name. done

>       * highlight the subpipeline node
>       * display inputs/outputs of the selected brick in the brick table below
> 

This should be done with last commit

>     * if the user click on a subpipeline node in the history window:
>       
>       * highlight the subpipeline node

done also

>       * display in the brick table a table that list inputs and outputs of the subpipeline --> to retrieve inputs/outputs files of the subpipeline, we will need to follow each input/output plug to find the first brick they are linked to, and retrieve in the database associated brick inputs/outputs files. Here again we are facing the same issue than above, we may find several corresponding bricks in the database...

I'm working on this that is a bit more complicated, and I will answer your questions soon @servoz

**Comment by LStruber** [comment]
In the meantime, I'll start looking at iterated pipelines

**Comment by servoz** [comment]
Since d79a636ee039dc09a7ecb3c89361cab487d57551 we cannot open anymore the history windows from a project, and Mia crashes.
Exception:
```
File "/data/Git_Projects/populse_mia/python/populse_mia/user_interface/data_browser/data_browser.py", line 2175, in show_brick_history
    self.brick_history_popup = PopUpShowHistory(
  File "/data/Git_Projects/populse_mia/python/populse_mia/user_interface/pop_ups.py", line 5008, in __init__
    self.node_selected(full_brick_name[0], pipeline.nodes[full_brick_name[0]])
KeyError: 'spatial_preprocessing_1_1'

```

**Comment by LStruber** [comment]
hmmm could you give a minimum steps to reproduce, I can't make it crash, I can open each history brick... (single bricks, bricks in subpipeline, bricks in imbricked subpipeline)

**Comment by servoz** [comment]
> What do you think to disable the node controller, as long as it is not well populated (and for now only use the brick row even if it is not the best way to read) ? It will let us the choice between populate the node controller or improve display of brick row mimicking controller

For me it's OK. We keep` 8) Node controller I/O should be populated --> let's leave that for later`, but until it's working, it's better to remove it. I think it's better to show only what works. To be able to reconnect easily and quickly, I think the best thing to do is to put in the code comment the number of this ticket and a small explanation.

What do you intend to do? If I remember correctly the `open node controller` is managed by capsul (pipeline_developer_view module). We don't want to change the capsul codes for this. Would you use _restricted_edition? (or something like that))

**Comment by LStruber** [comment]
`PipelineDeveloperView` was well designed and has an input called "allow_open_controller" that can be set to True or False !

**Comment by servoz** [comment]
> PipelineDeveloperView was well designed and has an input called "allow_open_controller" that can be set to True or False !

Ok, nice!

**Comment by servoz** [comment]
> hmmm could you give a minimum steps to reproduce, I can't make it crash, I can open each history brick... (single bricks, bricks in subpipeline, bricks in imbricked subpipeline)

Well, I'm just trying to open the history in the test project I use regularly. I will investigate...

**Comment by servoz** [comment]
Are you up to date with all populse?

**Comment by LStruber** [comment]
From what I see it may be a naming issue (between node_name and name and context_name etc.). I didn't try to rename pipeline. Did you rename it in your test pipeline ?

I should be up to date or almost, I did bv_maker this morning. I'll check

**Comment by servoz** [comment]
> Did you rename it in your test pipeline ?

No, it does not seem to me.

**Comment by LStruber** [comment]
Could you print "full_brick_name" l. 4970 ? (before any "pop")

**Comment by LStruber** [comment]
no since your full_brick_names should be "poped" l.4996

**Comment by servoz** [comment]
Yes changing by 
`self.node_selected(full_brick_name[0], pipeline.nodes[full_brick_name[-1]])` #l5008
works ... But I lose the highlighting of the nodes... ah and nothing change when I select a node in the history window ...

**Comment by LStruber** [comment]
and if you set -1 here, everything will go wrong with subpipelines, it should stay 0 here

**Comment by servoz** [comment]
I do not pass in l4996 since the pipeline have more than 2 nodes (if len(pipeline.nodes) == 2:)

**Comment by servoz** [comment]
did you test with the project obtained after running the preprocessing pipeline of mia_processes?

**Comment by LStruber** [comment]
I tried it with Smooth alone, Spatial preprocessing alone, Spatial preprocessing and Smooth on anat file before, and nested subpipelines (a spatial preproc with smooth I saved in user_processes and another smooth before).
It works in all cases on my side

**Comment by LStruber** [comment]
All bricks from mia_processe

**Comment by LStruber** [comment]
When I run a Spatial_Preprocessing alone, my pipeline have two nodes: one named '' and the other named 'spatial_preprocessing_1_1'

**Comment by LStruber** [comment]
Ah ! I managed to make it crash ! Once again, I exported all plugs before running pipeline. If I don't the pipeline is not the same, and I think you didn't export in your case too. I work on a fix

**Comment by servoz** [comment]
oh exactly i certainly took by chance the pipeline without export that i tested yesterday following the discussion with @denisri !!! !!!
phew, I thought I was going crazy ... 
On the other hand it reminds us that we still have to manage the case without export (which I didn't think about until yesterday and which is very useful for users!)

**Comment by LStruber** [comment]
wait... now it crashes in some other cases

**Comment by servoz** [comment]
ah and I just seen that you fixed a large part of the point 2 with 78894713f !!! Nice !!!.
So please let me the last part of this point ;-)

**Comment by servoz** [comment]
I leave you on the end of thefix and I go back on the end of the point 2
!

**Comment by LStruber** [comment]
ok it should be ok now, silly error (`is not` and `!=` is not equivalent in python !)

**Comment by LStruber** [comment]
Regarding point 2... If you feel hurt about me not letting you doing it, you can look (again) at UT that my changes seems to have broken !

**Comment by servoz** [comment]
> Regarding point 2... If you feel hurt about me not letting you doing it, you can look (again) at UT that my changes seems to have broken !

I don't feel hurt at all!!! the more important is that things are moving forward. It is the opposite that could hurt me :-)))
I will look at the UTs ...
I will look at the last point of point 2 (bad dispay of some parameters). It seems to be an old issue (that we didn't detect until yesterday) and that has nothing to do with PopUpShowHistory (well, I think I just started to look).

**Comment by LStruber** [comment]
1- Some pipeline cannot be reloaded from xml --> **@denisri work on a fix in Capsul. Exception should be handled in MIA**
2- Brick table issues in updating, column size and sometimes wrong data in some colums --> **@servoz**
3- Handle the case of bricks that are subpipelines --> **seems OK to me, @denisri can you check on your side**
4- Test and handle the case of iterated pipelines --> **TODO (@LStruber)**
5- Check if remove_orphan functions are still up to date with the new database, as well as (tightly related) the click on "cleanup obsolete items" button --> **TODO**
6- stdout contains too much information --> **open a ticket to do this later**
7- Close all windows when we close Mia (history, soma-workflow, etc.) --> **open a ticket to do this later**
8- Node controller inputs/outputs should be populated --> **for now, I disabled the node controller, open a ticket for dealing with that later**

**Comment by LStruber** [comment]
About point 4:
For now, when trying history with a iterated process, the pipeline view does not show (we see only the brick row). Indeed, there is an exception when reloading the pipeline from the xml string. Two solutions:
- Either we fix this in Capsul and allow the "reloading" of an iterated pipeline (I've no idea how difficult it could be)
- Or when saving the pipeline in initialization, we do not save the "iterated" pipeline, but the inner pipeline

I can implement solution 2 without too much trouble I think, but it will not handle the (rare) case where iterated pipeline is not alone (I don't even know if it's possible: create a pipeline, iterate it and then add some other nodes before and after). In this particular case nothing will be shown in the pipeline view.

**Comment by servoz** [comment]
I had never thought about this case. For me the current iteration in Mia only allows interate on the pipeline. I'm not sure if it would work if we add before or after something. It would be necessary to test, in fact I do not know...

**Comment by LStruber** [comment]
I pushed commits to save inner pipelines of iterated pipelines and to display it in the history window (the difficult part here was to associate the iterated pipeline process to the right brick in the database. Note that inputs/outputs of saved pipeline contains only the last iteration ! It may cause trouble when reactivating node controller)
I tested it with an iterated smooth, an iterated spatial preprocessing and an iterated smooth + spatial preprocessing (to test subpipelines). Please play with it on your side and tell me if you find errors

**Comment by LStruber** [comment]
I'll look at point 5 and the exception occuring when postprocessing a pipeline execution

**Comment by servoz** [comment]
Most of the issues for point 2 now seem to be fixed.

There are still two problems which certainly come from the initialisation (certainly a very old problem which has pass under the radar until now!):

a)The tuples are transformed into lists. I'll look into this now. I don't think it's from the controller because we see the right value in the controller's GUI  for the value of the `tissue` parameter of the segment spm brick (even if the V2 controller doesn't know how to handle this typing, but that's another problem). @sapetnioc, could this be due to populse_db? (which would transform the tuples into lists at the time of indexing or at the time of the query result). If not, we'll just have to make a small change in Mia (certainly at initialisation time. If it is, it will be more complicated.

b) The second problem concerns the `<undefined>` value. It seems that this value is indexed as "<undefined" in the database. Again, could this be from populs_db?

**Comment by servoz** [comment]
Wonderful the UTs are now down after my last change ....  I take it (maybe at the end of the point 2).

**Comment by servoz** [comment]
I just tested iterated spatial preprocessing and I see:
```
Iteration_pipeline has not run correctly:
'NoneType' object is not iterable

Traceback (most recent call last):
  File "/casa/home/Git_projects/populse_mia/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py", line 3030, in run
    self.pipeline_manager.postprocess_pipeline_execution(pipeline)
  File "/casa/home/Git_projects/populse_mia/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py", line 1919, in postprocess_pipeline_execution
    obsolete = self.project.update_data_history(outputs)
  File "/casa/home/Git_projects/populse_mia/python/populse_mia/data_manager/project.py", line 1049, in update_data_history
    obsolete.update(brick for brick in bricks if brick not in used)
TypeError: 'NoneType' object is not iterable

```
Apart from that, everything seems to be fine with this pipeline. All outputs are correctly created and the history window gives expected results. However, this pipeline is much simpler than @denisri . I think it's his pipeline that will be really revealing (crash test pipeline :-)))).

**Comment by LStruber** [comment]
yep that's the exception I was mentionning. On it !

**Comment by servoz** [comment]
> yep that's the exception I was mentionning. On it !

Oh yes !!!!!!

**Comment by LStruber** [comment]
And as I expected, playing with "garbage collect" button largely messes up the history ! I think that too much bricks are deleted in "cleanup_orphan" functions, we need to keep all bricks that are referenced in a pipeline

**Comment by servoz** [comment]
I found the culprit!
Its during initialisation [(PipelineManagerTab.init_pipeline()]((https://github.com/populse/populse_mia/blob/master/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py#L1670)) when indexing in db the values for COLLECTION_BRICK ([PipelineManagerTab._register_node_io_in_database(),](https://github.com/populse/populse_mia/blob/40ca06565d38afb994a3f1df81b3d646cd8cbfac/python/populse_mia/user_interface/pipeline_manager/pipeline_manager_tab.py#L376) we use this piece of code:

```
 for key in inputs:

            # filter Undefined / temp
            # this is an overhead since we convert to/from json, and it will
            # be converted again in the database. But the "default" function
            # for json is not in the database API.
            code = json.dumps(inputs[key], default=_serialize_tmp)
            inputs[key] = json.loads(code)

        for key in outputs:

            # filter Undefined / temp
            # this is an overhead since we convert to/from json, and it will
            # be converted again in the database.
            code = json.dumps(outputs[key], default=_serialize_tmp)
            outputs[key] = json.loads(code)
```
json.dumps() convert tuple to list and convert <undefined> to '"<undefined"' ... So the issue is may be not related to db indexing but  is strongly related ...
@sapetnioc, as you are the one of us who knows the database best, I am very interested in your point of view...

**Comment by servoz** [comment]
Ok for the point 2 I think the major part are now fixed. 
There's just one other thing wrong. The tuples get lost and become lists. To be as rigorous as possible, this is not good. However I can live with it (it doesn't fundamentally change the information in the context of viewing the history, in my opinion).
What do you think about it?

**Comment by servoz** [comment]
Ok only points 4 and 5 remain (the others are settled or in the process, or to be put in new tickets). According to your last posts you seem to be working on points 4 and 5 already @LStruber ? 
So I'm going to work on the TUs again, and then I can come back for points 4 and 5 if necessary.

**Comment by LStruber** [comment]
I finished working on point 4 this morning, and validated it on several pipelines, then it's settled also (unless @denisri or you find cases it does not work), and last commit should do the job for point 5, garbage collecting now do not mess up anymore the brick and history collections (and exception we discussed is gone)
I think we're closed to be able to close this ticket !

**Comment by LStruber** [comment]
And if history display is ok for everybody, we'll close #262 in the same time

**Comment by LStruber** [comment]
(I forgot to mention that if you find cases that do not work, please wait for next week to post it so I can spent a good week-end :D)

**Comment by servoz** [comment]
Great!

**Comment by servoz** [comment]
I have just redone some tests with my pipelines and it now works correctly.
I think @denisri  has a more complex test pipeline than mine. Let's wait and see what @denisri will say with a test on his pipeline. Then I propose to close this ticket and also #262.

Unless I am mistaken, this ticket still has items 1, 6, 7 and 8 to process. When we close this ticket, we can open 4 new tickets, one for each remaining item?

**Comment by servoz** [comment]
Working on another ticket (#264), I notice that there is still at least one problem.

To reproduce:

Take the Smooth brick (from mia_processes for example), setup the pipeline in simplified mode (no inputs/outputs nodes), select the in_files parameter and run the pipeline.
![Screenshot from 2022-04-19 12-41-14](https://user-images.githubusercontent.com/37933899/163986818-e5d7bc66-2311-47ff-a89e-a4a131978668.png)


The pipeline runs correctly (we get the smoothed image), but the name of the brick in the History tag is Smooth.main.
![Screenshot from 2022-04-19 12-42-40](https://user-images.githubusercontent.com/37933899/163987180-8d70c98d-6414-4355-8b50-40cfdb58f205.png)


As a consequence, if we click on the Smooth.main button, mia crashes with this exception:

  File "/casa/home/Git_projects/populse_mia/python/populse_mia/user_interface/data_browser/data_browser.py", line 2177, in show_brick_history
    self.data_browser.main_window)
  File "/casa/home/Git_projects/populse_mia/python/populse_mia/user_interface/pop_ups.py", line 5034, in __init__
    self.node_selected(full_brick_name[0], pipeline.nodes[full_brick_name[0]])
KeyError: 'Smooth'

**Comment by LStruber** [comment]
I handled this particular case in history display, but the name remains "Smooth.main" in the data browser. This name is "forced" by Capsul when the brick is alone and I didn't dig into it. I indeed think that this is tightly linked to #258

**Comment by servoz** [comment]
Great !!!!
[So we only have points 6, 7 and 8 left](https://github.com/populse/populse_mia/issues/263#issuecomment-1099516527). These are not priorities and we will be able to open 3 tickets to fix these points later.
If everyone agrees (everyone's tests are successful) we can close this ticket and also #162? 

@denisri:

- [what do you think of the name Smooth.main observed](https://github.com/populse/populse_mia/issues/263#issuecomment-1102493736). Do you confirm this is related to #258. ? or do we need to open a new ticket ?
- We can no longer run the Smooth brick from FSL (mia_processes). It worked perfectly well before (I don't know when it was still working!). Could this be related to the latest change in capsul?

**Comment by servoz** [comment]
> We can no longer run the Smooth brick from FSL (mia_processes). It worked perfectly well before (I don't know when it was still working!). Could this be related to the latest change in capsul?

It seems to come from Mia. I'm investigating...

**Comment by denisri** [comment]
> Great !!!! [So we only have points 6, 7 and 8 left](https://github.com/populse/populse_mia/issues/263#issuecomment-1099516527). These are not priorities and we will be able to open 3 tickets to fix these points later. If everyone agrees (everyone's tests are successful) we can close this ticket and also #262 ?

Yes, OK for me.

> @denisri:
> 
>     * [what do you think of the name Smooth.main observed](https://github.com/populse/populse_mia/issues/263#issuecomment-1102493736). Do you confirm this is related to #258. ? or do we need to open a new ticket ?

I guess it's because a single process is replaced with a small pipeline containing a single node. If it's ust for the name, the node has probably be named "main" hard-coded but could be changed to anything else (could be Smooth.Smooth here). Does it matter ?
Now the problem is more certainly the crash. The small pipeline is created on-the-fly and temporarily when needed, to it's maybe not the same instance which runs (or generates the workflow), and the one which is indexed in the database. Maybe this could cause a problem ?

>     * We can no longer run the Smooth brick from FSL (mia_processes). It worked perfectly well before (I don't know when it was still working!). Could this be related to the latest change in capsul?

No idea... I hope not, I haven't changed anything in the nipype or FSL stuff, but I'm not sure...
(I'm on vacation in a couple of hours, and coming back on may 2.)

**Comment by denisri** [comment]
I finally could create a "fake morphologist" pipeline that can be used without dependencies (other than Capsul) in tests. Look at:
https://github.com/populse/capsul/blob/master/capsul/pipeline/test/fake_morphologist/README.rst

**Comment by servoz** [comment]
> I  guess it's because a single process is replaced with a small pipeline containing a single node. If it's ust for the name, the node has probably be named "main" hard-coded but could be changed to anything else (could be Smooth.Smooth here). Does it matter ?
Now the problem is more certainly the crash. The small pipeline is created on-the-fly and temporarily when needed, to it's maybe not the same instance which runs (or generates the workflow), and the one which is indexed in the database. Maybe this could cause a problem ?

@LStruber fixed the crash yesterday. Concerning the name, I would have preferred it to be the same in simple mode or in inputs/outputs exported mode (simply for the sake of consistency for the user). But this is not a priority, we can come back to this later (there is already a ticket open on this subject, I think).

> No idea... I hope not, I haven't changed anything in the nipype or FSL stuff, but I'm not sure...
(I'm on vacation in a couple of hours, and coming back on may 2.)

I'm investigating this and I confess I'm a bit lost... If we take capsul commit 618f3fa0 and mia aeb88cb4 it works fine. If we take capsul 618f3fa0 (the same as before) and mia d640d339 (the next one from aeb88cb4), it doesn't work ... but the only difference between mia d640d339 and mia aeb88cb4 is the spm configuration ... @denisri can it could comes from somewhere else (soma?). Any ideas would be appreciated!!!

**Comment by servoz** [comment]
I have just opened 2 tickets for points 7 and 8 (#266 and #267, respectively).
Regarding point 6, I think there is no need to open a new ticket for this, the stdout is clean enough now!
